### ISSUE TITLE
Render using multiple threads.

### ISSUE THREAD

**Issue by ericsnowcurrently** [toolbox4minecraft/amidst#521]
Currently it looks like amidst renders one chunk (?) at a time in a single thread.  On a multi-core CPU rendering could be done more faster if multiple threads were used.

**Comment by moulins** [comment]
You're right, Amidst currently use a single thread to generate the biome data. This is because, until Minecraft 1.13, the Minecraft code for biome generation wasn't multithreadable.

Since the biome generation rewrite, the biome generation can be done in multiple threads, so Amidst could take advantage of that (but I haven't looked at it yet).

**Comment by ericsnowcurrently** [comment]
No worries.  It is a nice-to-have and I'd love to see this happen.  Hopefully it isn't much work.  That said, I don't want to impose on your time either way. :)

Also, thanks for the explanation.  I hadn't realized biome generation wasn't multi-threadable until recently.  That makes more sense now.

Anyway, thanks for a great tool!

**Comment by burgerguy** [comment]
> Until Minecraft 1.13, the Minecraft code for biome generation wasn't multithreadable.
Do you know what version exactly introduced this? I'm currently looking at possible implementations for this, but I can't seem to find the exact version that this was supported in.

**Comment by moulins** [comment]
The world generation was rewritten in 18w06a ;).

**Comment by burgerguy** [comment]
I saw your post on #395 and saw that the `getBiomeData` uses `LazyArea`s. This seems fine until you look at the `LazyArea` code and find that the `getValue` function (which is used in `getBiomeData`) is synchronized. This means that without some other method of getting the biome data, only one thread would be using it at a time, making it have no real performance benefit. The method that I tried and got the best performance (albeit worse than single threaded), was a work stealing thread pool when it calls for fragments to load. With one test on single core i got ~38 seconds zoomed out max, and the pool got ~54 seconds.

The getValue function for LazyArea:
```java
   public int getValue(int x, int z) {
      long i = this.func_202679_b(x, z);
      synchronized(this.cachedValues) {
         int j = this.cachedValues.get(i);
         if (j != Integer.MIN_VALUE) {
            return j;
         } else {
            int k = this.pixelTransformer.apply(x, z);
            this.cachedValues.put(i, k);
            if (this.cachedValues.size() > this.maxCacheSize) {
               for(int l = 0; l < this.maxCacheSize / 16; ++l) {
                  this.cachedValues.removeFirstInt();
               }
            }

            return k;
         }
      }
   }
```

I think that the approach that you were going for with MethodHandle would be much better for multithreading, as it might be able to run asynchronously. If I find anything else noteworthy I'll be sure to note it here too.

**Comment by burgerguy** [comment]
I changed the code a bit so instead it directly calls `getValue` from the layer's own `LazyArea`. This works so that we don't have to go from `int` to `Biome` to `int` to `short`, but we can instead get the `int`s directly so we only have to do one conversion to `short`. Sadly, this makes little to no speed improvement, and in my testing it was only around 1 second faster than the normal way. However, this makes it more robust so there are less points of failure. I tried multithreading again, but it was still just as slow because of the `synchronized` in the getValue. This also only works past 18w06a because it was just a replacement for IntCache iirc. Also @moulins, is there a better way to talk about optimizations than through github issues?

**Comment by magicus** [comment]
I started looking at Amidst performance, and how to get multithreaded fragment loading. I read the comment above about the locking in getValue(), and thought I should treat the issue as a Gordian knot, and just work around it using brute force. 

So I created multiple MinecraftInterfaces, each loading a copy of the mojang jar into a separate class loader, and dispatched to the different instances depending on which thread get to run getBiomeData(). While I'm a bit fuzzy on the details of class loading, I assumed this would make sure that each instance would get its own lock.

The results however was disappointing. At first I thought things worked fine, I saw multiple fragments being rendered simultaneously, and CPU usage capped at 100%. But when I started measuring wall clock time, I could not see any improvements. In fact, performance got significantly worse. :-(

I'm not sure what's causing this. Perhaps there is some other lock getting contended. Or my assumption about different class loaders having different objects to lock on is incorrect. Or there is some other issue that is causing the gains in performance to be lost to other factors.

**Comment by magicus** [comment]
Actually, it turned out my testing methodology was flawed in multiple ways. :-( 

First of all, I forgot that my laptop where I tested only was dual core, so I flooded the CPUs with too many threads. Lowering this to just two threads was much better. (I also want to try this on a machine with more cores to get an idea of the scaling.)

Secondly, I didn't account properly for the initial increase in overhead. One problem with my solution is that you need to call createWorld() once for every thread, and this is *not* cheap. I had to parallelize the creation and make the Fragment worker threads do it, because doing it in sequence from the main thread at startup took forever. But then I mistakenly included that initialization when comparing with my fix (but not when comparing without it). Of course the overhead added by this needs to be taken into consideration, but to understand the effect we need to compare apples to apples.

Third, I relied on visual inspection and a stop watch to determine the performance. Since we're talking about 10s of seconds of difference here, it was simple and good enough. However, it was hard to determine when the loading was done. I later noticed the "Fade fragments" setting, and when I turned it off, it became much clearer when loading was done, so I got more precision in my measurements.

And fourth, as I've said in #562, we spend a significant time on redrawing already rendered fragments. This means that the time spent on actually calculating the fragments is so low, so any difference this multithreading would give becomes so small it's difficult to track with any precision. So I increased the repaint tick to 1 second. This made the CPU loss to repainting minimal. The downside was that my precision with visual inspection and stop watch method decreased with in the worst case up to a second. On the other hand, now the difference between 1 and 2 threads became much more pronounced.

**Comment by magicus** [comment]
So, here is the new results:

When I measure just the time taken from the first image is painted until the last is painted, the time spent with one thread is 12 seconds, and with 2 threads it's 7 seconds, i.e. almost 50% time reduction. This is clearly in line with what could be expected.

However, due to the need to load the jar multiple times (takes a few extra seconds in total) and do the createWorld() multiple times (takes many seconds per world), the multithreaded solution actually starts out as slower, and only exceeds the single threaded version in performance after a while.

So I measured the startup time as well. This is from the start of the program, until the first initial map is completely loaded.

Startup time with Amidst @ master, unpatched: 36 seconds
Startup time with Amidst @ master, but with repaint tick at 1 second: 23 seconds

Using my patch, with two threads, repaint tick at 1 second, createWorld() called sequentially from main thread: 31 seconds
Using my patch, with two threads, repaint tick at 1 second, createWorld() called lazily in parallel from worker threads: 28 seconds

I have repeated some of the runs and gotten a variation of 1-2 seconds, so you need to take the times with a grain of salt.

Some things are clear, though:

1) Fixing the unnecessary repaint is the single biggest thing needed to increase performance of Amidst. Trying to do anything else related to performance before this is likely to have little effect, and be hard to measure.

2) Going the route I have explored seem to work fine to improve parallelism, but definitely comes at a cost of longer startup time. We need to make sure that there are no reasonable use cases in which this results in a regression rather than an improvement. (And maybe also see if createWorld() can be made faster.)

**Comment by burgerguy** [comment]
@magicus the redrawing of the fragments shouldn’t be a large deal considering (iirc) the rendering of fragments are done on a completely different thread.

**Comment by magicus** [comment]
@burgerguy I'm not sure you are correct. As my numbers above indicate, I got a massive speedup of the initial rendering by lowering the update frequency. If this is done in a separate thread, maybe the same thread also has the responsibility to calculate the fragments? I.e. in effect it's just everything else (GUI responses or whatever) that's in a separate thread?

And secondly, the frequent re-rendering is in itself a real problem. My computer spends an enormous amounts of CPU cycles on Amidst, just by having it running to show a static window. This makes it impossible to have it running in the background while I'm playing Minecraft, since it seriously lags the game. And, it's unnecessary adding to my power consumption and thus the global heating. ;-)

**Comment by magicus** [comment]
@burgerguy Try it yourself! Change line 86 in ThreadMaster to read 2000 instead of 20. Compile and run Amidst and be shocked by the results. The only drawback is the slow response from the UI, but the rendering is blistering fast.

**Comment by burgerguy** [comment]
@magicus Two additional single-thread executors are created on startup, one for displaying fragments and the other for creating fragments. These most likely run on different threads unless you have a low core count cpu. With 12 threads on my cpu, when amidst isn’t creating fragments it uses less than 5% of my cpu as a whole. However, amidst seems to like to crash when I’m running minecraft on a separate monitor and I alt+tab, without giving an error message or an error in the log. Also, when setting thread to display every second or two, it only performs about ~20% faster than when it updates 50 times a second. If you could post your specs that would give some insight on the issue.

**Comment by burgerguy** [comment]
Started working on a possible multi threaded fragment loader, but it seems that the image acceleration is much worse than single threaded, making performance about the same. I'll update if I find anything else worth sharing.

**Comment by magicus** [comment]
I'm sorry that I have not had enough time to work on amidst lately. I'm running on macOS, maybe the Java rendering is more problematic here than on other platforms? For me, amidst typically steal all cpu cycles even when doing nothing but sitting there showing an inert window.

**Comment by moulins** [comment]
> I'm sorry that I have not had enough time to work on amidst lately. I'm running on macOS, maybe the Java rendering is more problematic here than on other platforms? For me, amidst typically steal all cpu cycles even when doing nothing but sitting there showing an inert window.

On Mac OS X, Amidst enables OpenGL acceleration (see [here](https://github.com/toolbox4minecraft/amidst/blob/master/src/main/java/amidst/Amidst.java#L143)). Maybe disabling it will improve performance?

**Comment by burgerguy** [comment]
Just an update, my recent tries at multithreading are not producing great results. I tried to accomplish this by using a thread pool in `FragmentQueueProcessor.processQueues()` like this:
```java
public void processQueues() {
	Dimension dimension = dimensionSetting.get();
	updateLayerManager(dimension);
	processRecycleQueue();
	int threadCount;
	int maxSize = fwm.getMaximumPoolSize(); // fwm in this case is a custom thread pool
	while (loadingQueue.peek() != null) {
		if ((threadCount = fwm.getActiveCount()) < maxSize) {
			for (int i = 0; i < maxSize - threadCount; i++) {
				fwm.execute(() -> {
					loadFragment(dimension, loadingQueue.poll());
					updateLayerManager(dimensionSetting.get());
					processRecycleQueue();
				});
			}
		}
		try {
			Thread.sleep(1);
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
	}
	layerManager.clearInvalidatedLayers();
}
```

Originally, the fragments showed up messed up due to the threads overwriting eachother on the dataArray in LocalMinecraftInterface. I fixed this by instead assigning an array for each thread with this:
```java
private enum ThreadArrayMap {
	;

	private static volatile HashMap<Long, int[]> arrayMap = new HashMap<Long, int[]>();
    
	public static int[] getArray() {
		Long id = Thread.currentThread().getId();
		if (arrayMap.containsKey(id)) {
			return arrayMap.get(id);
		} else {
			return setArray(16384); // defaults to size of 128x128 fragment
		}
	}
    	
	public static int[] setArray(int size) {
		int[] array = new int[size];
		arrayMap.put(Thread.currentThread().getId(), array);
		return array;
	}
}
```

However, when I use this method for loading fragments I get such a small performance boost it's not even worth the extra processing power used, not to mention the random holes or switched around fragments. The biggest performance boost I could get was with 2 threads working and it was about 2 seconds faster rendering a fully zoomed out window.

Any other ideas for how to go about this?

**Comment by moulins** [comment]
I think you're hitting some internal synchronization in the Minecraft code; you probably need a separate `BiomeProvider` for each thread.

**Comment by burgerguy** [comment]
In my implementation that I’ve been multithreading with I specifically bypass the synchronized in LazyArea. Also I’m hesitant to create a lot of BiomeProviders / BiomeSources because it probably creates a lot of time on startup. I’ll see what I can do with the current implementation when I get the time to, but it seems like it’ll be hard to manage all of the multithreading in the MinecraftInterface alone. I’ve been able to make the getBiomeData method thread safe by giving each thread its own temporary biomedata array without using any synchronized methods (that I know of). When I get the chance I’ll profile it.

**Comment by burgerguy** [comment]
Ok so I’ve created a very early version of multithreading here (#700) and it runs extremely fast compared to the current version. I hope the bugs will be easy to iron out because this is a really needed change for amidst.

**Comment by magicus** [comment]
@burgerguy Going back to your post from January (yay, backlog!), I was more or less assuming that the issue was the MacOS rendering pipeline. But now I wanted to verify it, so I installed Amidst on my linux machine as well. The same story. When Amidst has stopped calculating new fragments, the CPU is still consuming lot of CPU. Xorg goes up to 40-45% constantly. Since I have more cores on this machine, it's not as noticeable as on my weaker MacBook, but it's still annoying. As soon as I minimize the window, the Xorg CPU drops down to undetectable.

**Comment by magicus** [comment]
@moulins I have not tested trying to disable OpenGL. (Not sure how that could even help things..?) I did however try to build the new project Lanai OpenJDK, which adds macOS rendering using the Metal pipeline. It did work just as well (despite being experimental still), but it had similar performance characteristics.
